includes/traits/llms-trait-earned-engagement-meta-box.php

```diff
@@ -0,0 +1,177 @@
+<?php
+/**
+ * LifterLMS Eearned Engagements (Certificate/Achievement) Meta Box trait.
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LifterLMS Eearned Engagements (Certificate/Achievement) Meta Box trait.
+ *
+ * **This trait should only be used by classes that extend from the {@see LLMS_Admin_Metabox} class.**
+ *
+ * @since [version]
+ */
+trait LLMS_Trait_Earned_Engagement_Meta_Box {
+
+	/**
+	 * Allowed post types.
+	 *
+	 * @since [version]
+	 *
+	 * @var string[]
+	 */
+	private $allowed_post_types = array(
+		'llms_my_achievement' => array(
+			'model'                => 'LLMS_User_Achievement',
+			'user_postmeta_prefix' => '_achievement',
+		),
+		'llms_my_certificate' => array(
+			'model'                => 'LLMS_User_Certificate',
+			'user_postmeta_prefix' => '_certificate',
+		),
+	);
+
+	/**
+	 * Add earned engagement fields.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $fields  Array of metabox fields.
+	 * @return array
+	 */
+	protected function add_earned_engagement_fields( $fields = array() ) {
+
+		$post_type = get_post_type();
+		if ( ! array_key_exists( $post_type, $this->allowed_post_types ) ) {
+			return $fields;
+		}
+
+		$student = ! empty( $_GET['sid'] ) ? llms_filter_input( INPUT_GET, 'sid', FILTER_SANITIZE_NUMBER_INT ) : false; // phpcs:ignore
+		$student = empty( $student ) ? ( new $this->allowed_post_types[ $post_type ]['model']( $this->post->ID ) )->get_user_id() : $student;
+
+		if ( empty( $student ) ) {
+			$fields[] = array(
+				'allow_null'      => false,
+				'class'           => 'llms-select2-student',
+				'data_attributes' => array(
+					'allow_clear' => false,
+					'placeholder' => __( 'Select a Student', 'lifterlms' ),
+				),
+				'id'              => $this->prefix . 'student',
+				'label'           => __( 'Select a Student', 'lifterlms' ),
+				'type'            => 'select',
+			);
+		} else {
+			array_unshift(
+				$fields,
+				array(
+					'id'    => $this->prefix . 'student',
+					'type'  => 'hidden',
+					'value' => $student,
+				)
+			);
+		}
+
+		return $fields;
+	}
+
+	/**
+	 * Save a metabox field.
+	 *
+	 * @since [version]
+	 *
+	 * @param int   $post_id WP_Post ID.
+	 * @param array $field   Metabox field array.
+	 * @return boolean
+	 */
+	protected function save_field( $post_id, $field ) {
+
+		if ( $this->prefix . 'student' === $field['id'] && isset( $_POST[ $field['id'] ] ) ) { //phpcs:ignore -- nonce already verified in `LLMS_Admin_Metabox::save`.
+			$this->log_earned_engament( llms_filter_input( INPUT_POST, $field['id'], FILTER_SANITIZE_NUMBER_INT ), $post_id );
+		}
+
+		/**
+		 * Skip saving _llms_student field, only used to award an engagement, it's not a post field.
+		 */
+		if ( isset( $field['id'] ) && $this->prefix . 'student' === $field['id'] ) {
```

Comment:
On my branch I'm switching to save the user id to the `post_author` field. This will make querying data more efficient in the long run (instead of relying on our congested user postmeta table. We need a data migration and I plan to continue storing the postmeta data for backwards compatiblity reasons though.

Reply:
Alright, this means that I have to save that field value into the `post_author` on creation... ?

Reply:
Yes :-/